src/video_core/texture_cache/image.cpp

```diff
@@ -179,52 +180,130 @@ Image::Image(const Vulkan::Instance& instance_, Vulkan::Scheduler& scheduler_,
                           info.guest_size_bytes);
 }
 
-void Image::Transit(vk::ImageLayout dst_layout, vk::Flags<vk::AccessFlagBits> dst_mask,
-                    vk::CommandBuffer cmdbuf) {
-    if (dst_layout == layout && dst_mask == access_mask) {
-        return;
+boost::container::small_vector<vk::ImageMemoryBarrier2, 32> Image::GetBarriers(
+    vk::ImageLayout dst_layout, vk::Flags<vk::AccessFlagBits2> dst_mask,
+    vk::PipelineStageFlags2 dst_stage, std::optional<SubresourceRange> subres_range) {
+    const bool needs_partial_transition =
+        subres_range &&
+        (subres_range->base != SubresourceBase{} || subres_range->extent != info.resources);
+    const bool partially_transited = !subresource_states.empty();
+
+    boost::container::small_vector<vk::ImageMemoryBarrier2, 32> barriers{};
+    if (needs_partial_transition || partially_transited) {
+        if (!partially_transited) {
+            subresource_states.resize(info.resources.levels * info.resources.layers);
+            std::fill(subresource_states.begin(), subresource_states.end(), last_state);
+        }
+
+        // In case of partial transition, we need to change the specified subresources only.
+        // Otherwise all subresources need to be set to the same state so we can use a full
+        // resource transition for the next time.
+        const auto mips =
+            needs_partial_transition
+                ? std::ranges::views::iota(subres_range->base.level,
+                                           subres_range->base.level + subres_range->extent.levels)
+                : std::views::iota(0u, info.resources.levels);
+        const auto layers =
+            needs_partial_transition
+                ? std::ranges::views::iota(subres_range->base.layer,
+                                           subres_range->base.layer + subres_range->extent.layers)
+                : std::views::iota(0u, info.resources.layers);
+
+        for (u32 mip : mips) {
+            for (u32 layer : layers) {
+                // NOTE: these loops may produce a lot of small barriers.
+                // If this becomes a problem, we can optimize it by merging adjacent barriers.
+                auto& state = subresource_states[mip * info.resources.layers + layer];
+
+                if (state.layout != dst_layout || state.access_mask != dst_mask) {
```

Comment:
Got some memory corruption with this PR pulled, took a look with ASAN as well as adding some debugging and `mip * info.resources.layers + layer` is exceeding the size of `subresource_states`. Game is CUSA16429.

Reply:
The crash is happening because depth overlap resolving function doesn't support images with different numbers of slices yet. I'll add few asserts to track this case easier and will fix it a bit later